### ISSUE TITLE
Multilabel classification after detection

### ISSUE THREAD

**Issue by abhigoku10** [feiyuhuahuo/Yolact_minimal#42]
@feiyuhuahuo  thanks for you wonderful work and sharing the code based , i want to add a multiclass label classification for person which gives the gender and age , i want to add this to the yolact architecture as a head any suggestions 
Thansk in adavance

**Comment by feiyuhuahuo** [comment]
![image](https://user-images.githubusercontent.com/32631344/132292029-6920aa74-6930-47ab-bfa9-5a695dace5d6.png)
You can simply add additional two branches in the prediction head. Or add another head to predict gender, age. These two ways lead to different parameter numbers. Less shared parameters mean lower precision but higher speed, I think. There's another thing need to pay attention. If you just get the corresponding gender and age related to the NMSed box, they may not be the best. Some other optimizations need to be done. You can consult to the mask-scoring module in Mask R-CNN.

**Comment by abhigoku10** [comment]
@feiyuhuahuo  thanks for the response  the two which you mentioned is the ideal approach but the problem is with the NMS module since we have to apply to NMSed box it is as good as having detection + gender classification  instead of person detection || gender classification .. Do u think if we have anchor free approach it will help us to add gender and age classifier easily ???

**Comment by feiyuhuahuo** [comment]
I can‘t get you clearly. Do you mean you want to add another module after Yolact to do gender classification and age prediction? If so, maybe training another classification network works. But I don't know how to do classification directly on NMSed boxes. NMS has nothing to do with gradient, so such training is not achievable. I am also not sure whether what you mentioned has anything to do with anchor free approach. The common way is training two networks.

**Comment by abhigoku10** [comment]
@feiyuhuahuo my explanation is , if we want a seperate classifier within the network for gender and age its nto possibile because after detection we have to add that module and we have to check for person box since many boxes generates before nms  then adding classification module is computationaly high , so we have anchor free approach then adding additional classifier becomes easy right

**Comment by feiyuhuahuo** [comment]
It will not add too much computation cost if you add few convolution layers in the gender and age branch. Because the predicition head is shared across the FPN layers. And computation cost has nothing to do with anchor free, anchor free does not influence the boxes number to predict.

**Comment by abhigoku10** [comment]
@feiyuhuahuo  wokay thanks for your response , i followed your suggestion and created a gender classification for now , below are the queries 
1. In your custom.json i created a seperate field as "gender":Male , based on the image changed to Female or unknown so there are 3 classes which i am representing by 0,1,2 in my input data , and have made changes accordingly in the coco.py but in the future if we want to have multi attributes for the one label then how the input data should be represented 
2. In box_utils.py  in match function you have defined  "conf = class_gt[anchor_max_i] + 1  # the class of the max IoU gt box for each anchor
    conf[each_anchor_max < cfg.pos_iou_thre] = -1  # label as neutral
    conf[each_anchor_max < cfg.neg_iou_thre] = 0  # label as background" in my statement if i donot want 0as background then wat should i do since have this i am getting error "Target 2 outof bounds" for F.cross entropy 

can you share your thoughts on this

**Comment by feiyuhuahuo** [comment]
1.  Do you mean how the input data be organized in the coco.py? There's no special way for inputting data. If only the corresponding label can be input correctly when computing loss, the network can be trained successfully.
2.  Why you don't want 0 as background? Background is always needed for training an object detection program.

**Comment by feiyuhuahuo** [comment]
1.  Do you mean how the input data be organized in the coco.py? There's no special way for inputting data. If only the corresponding label can be input correctly when computing loss, the network can be trained successfully.
2.  Why you don't want 0 as background? Background is always needed for training an object detection program.

**Comment by abhigoku10** [comment]
@feiyuhuahuo Pt2: i want background for only object detection but i am doing a multitask network like gender where background is not required 
Pt1: thanks for response

**Comment by feiyuhuahuo** [comment]
Background is used for  classification for negative sample. When coming to issues like box regression, only positive samples are concerned. Gender prediction can also adopt such a way.

**Comment by abhigoku10** [comment]
@feiyuhuahuo  thanks for response , suggestion makes sense , now one query if i try to combine gender and age classification into single attribute classification then below are the queries i am having assuming there is male person in an image 
1. the annotation should be  label list = background , male , female ,age lessthan30 , age greaterthan 30 so should the annotation be [01001] or [14] ? which format do u suggest 
2. Based on the above annotation data what kind of loss do u suggest since we need to get predication as person-male-age greater than30 ideal

**Comment by feiyuhuahuo** [comment]
background is not needed, it's only for category classification. I suggest you to learn how object detection programs are trained and inferenced, especially how the training samples are determined and how the losses are adopted. When training, the corresponding gender and age attributes of positive samples are trained, when inference, the corresponding gender and age attributes of NMSed box are adopted. Thus annotation should be [1, 0, 0, 1], one-hot style of course. And BCE loss for classfication.

**Comment by abhigoku10** [comment]
@feiyuhuahuo  thanks for the responses i was able to add an additional head for gender and below results i am getting 
![MicrosoftTeams-image](https://user-images.githubusercontent.com/22094053/140465121-e3f4aa56-ffa2-42c9-adb4-5e55971cf065.png)


Now planning to add the age classification and check the results !!!
